File: packages/ui/db/entity/Boulder.ts
import { Area, Commentable, Coordinate, Photoable, Polygon, Route } from "@/db";
import { cascadeManyToOne, cascadeOneToMany } from "@/db/cascadeOptions";
import BaseColumnSchemaPart from "@/db/entity/BaseColumnSchemaPart";
import CoordinateSchema from "@/db/entity/Coordinate";
import { type IBoulder } from "models";
import { EntitySchema } from "typeorm";

export type Boulder = IBoulder & {
  polygon?: Polygon;
  area?: Area;
  routes?: Route[];
  coordinates: Coordinate;
  commentable?: Commentable;
  photoable?: Photoable;
};

const BoulderSchema = new EntitySchema<Boulder>({
  name: "boulder",
  columns: {
    ...BaseColumnSchemaPart,
    name: {
      type: String,
    },
    description: {
      type: String,
      nullable: true,
    },
  },
  relations: {
    area: {
      type: "many-to-one",
      target: "area",
      ...cascadeManyToOne,
    },
    routes: {
      type: "one-to-many",
      inverseSide: "boulder",
      target: "route",
      ...cascadeOneToMany,
    },
    polygon: {
      type: "one-to-one",
      target: "polygon",
      joinColumn: true,
      onDelete: "SET NULL",
      cascade: ["insert", "update"],
    },
    commentable: {
      type: "one-to-one",
      nullable: true,
      target: "commentable",
      onDelete: "SET NULL",
      joinColumn: true,
    },
    photoable: {
      type: "one-to-one",
      nullable: true,
      target: "photoable",
      onDelete: "SET NULL",
      joinColumn: true,
    },
  },
  embeddeds: {
    coordinates: {
      schema: CoordinateSchema,
      prefix: "coordinates_",
    },
  },
});

export default BoulderSchema;

File: go-server/models/boulder.go
package models

type Boulder struct {
	ID          *int         `json:"id,omitzero"`
	Name        string       `json:"name"`
	Description *string      `json:"description,omitzero"`
	Coordinates Coordinate   `json:"coordinates"`
	Area        *Area        `json:"area,omitzero"`
	Routes      []Route      `json:"routes,omitzero"`
	Polygon     *Polygon     `json:"polygon,omitzero"`
	Commentable *Commentable `json:"commentable,omitzero"`
	Photoable   *Photoable   `json:"photoable,omitzero"`
}

func (b Boulder) IsZero() bool {
	return b.ID == nil && b.Name == "" && b.Description == nil && b.Coordinates.IsZero() && b.Area == nil &&
		len(b.Routes) == 0 && b.Polygon == nil && b.Commentable == nil && b.Photoable == nil
}
